### PR TITLE
Bug/1255 Duplicating a objective does not bring all properties of the KeyResults

### PR DESCRIPTION
**Pull request by ManuelMoeri**

### CODE REVIEW

File: backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java [hunk 1]
```diff
@@ -107,64 +115,65 @@ public Objective createEntity(Objective objective, AuthorizationUser authorizati
...
                 .withDescription(keyResult.getDescription()) //
                 .withOwner(keyResult.getOwner()) //
-                .withCommitZone("-") //
-                .withTargetZone("-") //
-                .withStretchZone("-") //
+                .withCommitZone(((KeyResultOrdinal) keyResult).getCommitZone()) //
+                .withTargetZone(((KeyResultOrdinal) keyResult).getTargetZone()) //
+                .withStretchZone(((KeyResultOrdinal) keyResult).getStretchZone()) //
                 .build();
     }
```
Comment (MasterEvarior) [inline backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java]: That should belong to the `KeyResultBusinessService`. For the `ObjectiveBusinessService` the duplication of a key result should be abstracted.

I know they were here before, so it is not your fault but please move them anyway.

File: backend/src/main/java/ch/puzzle/okr/service/persistence/ActionPersistenceService.java [hunk 2]
```diff
@@ -22,4 +23,8 @@ public String getModelName() {
     public List<Action> getActionsByKeyResultIdOrderByPriorityAsc(Long keyResultId) {
         return getRepository().getActionsByKeyResultIdOrderByPriorityAsc(keyResultId);
     }
+
+    public void saveDuplicatedActions(List<Action> actions) {
```
Comment (MasterEvarior) [inline backend/src/main/java/ch/puzzle/okr/service/persistence/ActionPersistenceService.java:26]: This method has not really anything to do with duplication except that it is the first use case we needed it for.
Either:
- rename it to something more generic like `saveActions`
- replace it all together with a simple `.forEach` at the place you call this method

If we leave this method in place, it could also be argued that it can be generified in the `PersistenceBase` but I'd leave this for now as it is not necessary as of now.

File: backend/src/main/java/ch/puzzle/okr/service/business/ActionBusinessService.java [hunk 3]
```diff
@@ -74,4 +76,25 @@ public void deleteEntityById(Long id) {
     public void deleteEntitiesByKeyResultId(Long keyResultId) {
         getActionsByKeyResultId(keyResultId).forEach(action -> deleteEntityById(action.getId()));
     }
+
+    public List<Action> createDuplicateOfActions(KeyResult oldKeyResult, KeyResult newKeyResult) {
```
Comment (MasterEvarior) [inline backend/src/main/java/ch/puzzle/okr/service/business/ActionBusinessService.java:80]:
```suggestion
    public List<Action> createDuplicates(KeyResult oldKeyResult, KeyResult newKeyResult) {
```
I would rename this, simply because the `Actions` part is already heavily implied by it being in the `ActionBusinessService` :smiley:

File: frontend/src/app/services/objective.service.ts [hunk 4]
```diff
@@ -33,23 +31,10 @@ export class ObjectiveService {
...
-        version: number;
-        createdOn: Date | null | undefined;
-        objective: Objective;
-      }[];
-      objective: any;
-    }): Observable<Objective> {
-    return this.httpClient.post<Objective>(`/api/v2/objectives/${objectiveId}`, duplicateObjectiveDto);
+  duplicateObjective(duplicateObjectiveDto: {
+    keyResultIds: (number | undefined)[];
+    objective: any;
```
Comment (MasterEvarior) [inline frontend/src/app/services/objective.service.ts:36]: This is nice and way simpler than before. Good job :+1:

File: frontend/cypress/e2e/duplicate-objective.cy.ts [hunk 5]
```diff
@@ -96,6 +96,24 @@ describe('functionality of duplicating objectives and their belonging key-result
...
+    cy.contains('Metrisch');
+    cy.contains('Paco Eggimann');
+    cy.contains('GJ 24/25-Q3');
+    cy.contains('Action Plan')
+      .then(() => {
+        cy.contains('Neue Pflanzen');
+        cy.contains('Ein Buch');
+        cy.contains('Mehr Getränke');
+      });
+  });
```
Comment (MasterEvarior) [inline frontend/cypress/e2e/duplicate-objective.cy.ts:116]: If I am not mistaken this only tests the metric keyresult. There should also be one for ordinal if possible. This will save us fixing it again in 3-6 months :smiley:

File: backend/src/test/java/ch/puzzle/okr/service/persistence/ActionPersistenceServiceIT.java [hunk 6]
```diff
@@ -119,7 +121,7 @@ void shouldThrowExceptionWhenSaveIsCalledOnAlreadyUpdatedAction() {
     @Test
     void shouldReturnListOfAllActionsWhenFindAllIsCalled() {
         List<Action> actions = actionPersistenceService.findAll();
-
+        System.out.println(actions.getFirst());
```
Comment (MasterEvarior) [inline backend/src/test/java/ch/puzzle/okr/service/persistence/ActionPersistenceServiceIT.java:121]:
```suggestion
```
Yeet :airplane:

**Review by MasterEvarior — CHANGES_REQUESTED** [review]
Overall it looks already quite good. There are a couple of minor things and also the the organization of the code that should be improved.

File: backend/src/main/java/ch/puzzle/okr/service/persistence/ActionPersistenceService.java (hunk 2 above)
Comment (ManuelMoeri) [reply backend/src/main/java/ch/puzzle/okr/service/persistence/ActionPersistenceService.java:26]: Has been done

File: backend/src/test/java/ch/puzzle/okr/controller/ObjectiveControllerIT.java [hunk 7]
```diff
@@ -341,16 +341,13 @@ void shouldThrowExceptionWhenObjectiveWithIdCantBeFoundWhileDeleting() throws Ex
...
     @Test
     void shouldReturnIsCreatedWhenObjectiveWasDuplicated() throws Exception {
-        BDDMockito
-                .given(objectiveAuthorizationService.duplicateEntity(anyLong(), any(), anyList()))
-                .willReturn(objective1);
         BDDMockito.given(keyResultMapper.toDto(any(KeyResultMetric.class), any())).willReturn(keyResultMetricDto);
         BDDMockito.given(keyResultMapper.toDto(any(KeyResultOrdinal.class), any())).willReturn(keyResultOrdinalDto);
         BDDMockito.given(objectiveAuthorizationService.getAuthorizationService()).willReturn(authorizationService);
-        BDDMockito.given(objectiveMapper.toDto(objective1)).willReturn(objective1Dto);
+        BDDMockito.given(objectiveMapper.toDto(any())).willReturn(objective1Dto);
```
Comment (peggimann) [inline backend/src/test/java/ch/puzzle/okr/controller/ObjectiveControllerIT.java:350]: be careful using any here, could lead to unexpected behavior. Generally it's considered to be best practice to mock as precise as possible.

File: backend/src/main/java/ch/puzzle/okr/models/keyresult/KeyResult.java [hunk 8]
```diff
@@ -256,6 +255,11 @@ public T withModifiedOn(LocalDateTime modifiedOn) {
             return (T) this;
         }
 
+        public T withActionList(List<Action> actionList) {
+            this.actionList = actionList;
```
Comment (peggimann) [inline backend/src/main/java/ch/puzzle/okr/models/keyresult/KeyResult.java:259]: i think this is no longer needed to be done.

File: backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java [hunk 9]
```diff
@@ -26,15 +28,21 @@ public class ObjectiveBusinessService implements BusinessServiceInterface<Long,
...
+    private final ActionPersistenceService actionPersistenceService;
+    private final ActionBusinessService actionBusinessService;
 
     public ObjectiveBusinessService(@Lazy KeyResultBusinessService keyResultBusinessService,
                                     ObjectiveValidationService validator,
                                     ObjectivePersistenceService objectivePersistenceService,
-                                    CompletedBusinessService completedBusinessService) {
+                                    CompletedBusinessService completedBusinessService,
+                                    ActionPersistenceService actionPersistenceService,
+                                    ActionBusinessService actionBusinessService) {
```
Comment (peggimann) [inline backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java:39]: action business service should be enough there should be no need to use action persisstence service here

File: backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java [hunk 10]
```diff
@@ -107,64 +115,70 @@ public Objective createEntity(Objective objective, AuthorizationUser authorizati
...
+            newKeyResult = keyResultBusinessService.createEntity(keyResultMetric, authorizationUser);
         } else if (keyResult.getKeyResultType().equals(KEY_RESULT_TYPE_ORDINAL)) {
             KeyResult keyResultOrdinal = makeCopyOfKeyResultOrdinal(keyResult, duplicatedObjective);
-            keyResultBusinessService.createEntity(keyResultOrdinal, authorizationUser);
+            newKeyResult = keyResultBusinessService.createEntity(keyResultOrdinal, authorizationUser);
+        } else {
+            throw new UnsupportedOperationException("Unsupported KeyResultType: " + keyResult.getKeyResultType());
         }
+
+        List<Action> copiedActions = actionBusinessService.createDuplicateOfActions(keyResult, newKeyResult);
```
Comment (peggimann) [inline backend/src/main/java/ch/puzzle/okr/service/business/ObjectiveBusinessService.java:155]: save the entity in actionbusinessservice rather than in object business service

**Review by peggimann — DISMISSED** [review]

File: backend/src/main/java/ch/puzzle/okr/service/business/KeyResultBusinessService.java [hunk 11]
```diff
@@ -137,4 +143,49 @@ private boolean isKeyResultTypeChangeable(Long id) {
...
+
+        if (keyResult.getKeyResultType().equals(KEY_RESULT_TYPE_METRIC)) {
+            KeyResult keyResultMetric = makeCopyOfKeyResultMetric(keyResult, duplicatedObjective);
+            newKeyResult = createEntity(keyResultMetric, authorizationUser);
+        } else if (keyResult.getKeyResultType().equals(KEY_RESULT_TYPE_ORDINAL)) {
+            KeyResult keyResultOrdinal = makeCopyOfKeyResultOrdinal(keyResult, duplicatedObjective);
+            newKeyResult = createEntity(keyResultOrdinal, authorizationUser);
+        } else {
+            throw new UnsupportedOperationException("Unsupported KeyResultType: " + keyResult.getKeyResultType());
+        }
```
Comment (MasterEvarior) [inline backend/src/main/java/ch/puzzle/okr/service/business/KeyResultBusinessService.java:160]:
```suggestion
                KeyResult newKeyResult = switch (keyResult.getKeyResultType()){
            case KEY_RESULT_TYPE_METRIC -> makeCopyOfKeyResultMetric(keyResult, duplicatedObjective);
            case KEY_RESULT_TYPE_ORDINAL -> makeCopyOfKeyResultOrdinal(keyResult, duplicatedObjective);
            default -> throw new UnsupportedOperationException("Unsupported KeyResultType: " + keyResult.getKeyResultType());
        };
        newKeyResult = createEntity(newKeyResult, authorizationUser);
        
```
This could be simplified somewhat by using the newish switch-statement instead of an if-else chain.

File: backend/src/test/java/ch/puzzle/okr/service/business/ActionBusinessServiceTest.java [hunk 12]
```diff
@@ -170,4 +170,64 @@ void shouldDeleteAction() {
         actionBusinessService.deleteEntityById(action1.getId());
         verify(actionPersistenceService, times(1)).deleteById(action1.getId());
     }
+
+    @DisplayName("Should return a empty list when no actions are defined")
+    @Test
+    void shouldReturnEmptyListWhenActionsAreEmpty() {
```
Comment (MasterEvarior) [inline backend/src/test/java/ch/puzzle/okr/service/business/ActionBusinessServiceTest.java:176]: This should also test for `null`, would be a shame to not use that nice `if-clause` :smile: 
You can simply do this by making this a `@ParameterizedTest` with a `@NullAndEmptySource`.

File: frontend/cypress/e2e/duplicate-objective.cy.ts [hunk 13]
```diff
@@ -96,6 +98,73 @@ describe('functionality of duplicating objectives and their belonging key-result
     cy.contains(duplicatedTitle)
       .should('not.exist');
   });
+
+  it.only('should duplicate all attributes of the corresponding keyResults as well', () => {
```
Comment (MasterEvarior) [inline frontend/cypress/e2e/duplicate-objective.cy.ts]: Test is good, `only` not so much.

**Review by MasterEvarior — CHANGES_REQUESTED** [review]

File: backend/src/test/java/ch/puzzle/okr/service/business/ActionBusinessServiceTest.java [hunk 14]
```diff
@@ -170,4 +171,64 @@ void shouldDeleteAction() {
         actionBusinessService.deleteEntityById(action1.getId());
         verify(actionPersistenceService, times(1)).deleteById(action1.getId());
     }
+
+    @ParameterizedTest(name = "Should return a empty list when actions are null or empty")
```
Comment (MasterEvarior) [inline backend/src/test/java/ch/puzzle/okr/service/business/ActionBusinessServiceTest.java:175]:
```suggestion
    @ParameterizedTest(name = "Should return a empty list when actions are {0}")
```
Makes this a tiny bit nicer 
image

**Review by MasterEvarior — APPROVED** [review]
:tada: :tada:
